File: rupypy/modules/comparable.py
from rupypy.module import Module, ModuleDef


class Comparable(Module):
    moduledef = ModuleDef("Comparable", filepath=__file__)

    moduledef.app_method("""
    def > other
        return (self <=> other) > 0
    end

    def < other
        return (self <=> other) < 0
    end

    def >= other
        return !((self <=> other) < 0)
    end

    def <= other
        return !((self <=> other) > 0)
    end

    def == other
        return (self <=> other) == 0
    end

    def between? min, max
        return self >= min && self <= max
    end
    """)

File: lib-topaz/comparable.rb
module Comparable
  def >(other)
    return Topaz.compare(self, other) > 0
  end

  def <(other)
    return Topaz.compare(self, other) < 0
  end

  def >=(other)
    return Topaz.compare(self, other) >= 0
  end

  def <=(other)
    return Topaz.compare(self, other) <= 0
  end

  def ==(other)
    begin
      compared = (self <=> other)
    rescue StandardError
      return false
    end

    return compared == 0
  end

  def between?(min, max)
    return self >= min && self <= max
  end
end
